File: packages/cpp/src/obj_export.cpp
#include "openskp/obj_export.hpp"

#include <fstream>
#include <iomanip>
#include <sstream>
#include <stdexcept>

namespace openskp {

std::string to_obj(const Scene& scene) {
  std::ostringstream ss;
  ss.imbue(std::locale::classic());
  ss << "# OpenSKP OBJ Export\n";
  ss << "# Primitives: " << scene.glb_primitives.size() << "\n\n";

  std::uint32_t vert_offset = 1;  // OBJ indices are 1-based
  for (const auto& prim : scene.glb_primitives) {
    ss << "o " << prim.geom_name << "\n";

    std::size_t vert_count = prim.positions.size() / 3;
    for (std::size_t i = 0; i < vert_count; ++i) {
      ss << "v " << std::fixed << std::setprecision(6) << prim.positions[i * 3] << " "
         << prim.positions[i * 3 + 1] << " " << prim.positions[i * 3 + 2] << "\n";
    }

    std::size_t tri_count = prim.indices.size() / 3;
    for (std::size_t i = 0; i < tri_count; ++i) {
      std::uint32_t i0 = prim.indices[i * 3] + vert_offset;
      std::uint32_t i1 = prim.indices[i * 3 + 1] + vert_offset;
      std::uint32_t i2 = prim.indices[i * 3 + 2] + vert_offset;
      ss << "f " << i0 << " " << i1 << " " << i2 << "\n";
    }

    vert_offset += static_cast<std::uint32_t>(vert_count);
    ss << "\n";
  }

  return ss.str();
}

void export_obj(const Scene& scene, const std::filesystem::path& path) {
  auto parent = path.parent_path();
  if (!parent.empty() && !std::filesystem::exists(parent)) {
    std::filesystem::create_directories(parent);
  }
  std::ofstream out(path);
  if (!out) {
    throw std::runtime_error("Cannot open OBJ output file: " + path.string());
  }
  out << to_obj(scene);
  if (!out) {
    throw std::runtime_error("Failed to write OBJ output file: " + path.string());
  }
}

}  // namespace openskp

File: packages/dart/lib/src/obj_export.dart
import 'dart:io';

import 'scene.dart';

/// Convert a baked [Scene] into Wavefront OBJ text representation.
String toObj(Scene scene) {
  final buffer = StringBuffer();
  buffer.writeln('# OpenSKP OBJ Export');
  buffer.writeln('# Primitives: ${scene.glbPrimitives.length}');
  buffer.writeln();

  var vertOffset = 1; // OBJ indices are 1-based
  for (final prim in scene.glbPrimitives) {
    buffer.writeln('o ${prim.geomName}');

    final vertCount = prim.positions.length ~/ 3;
    for (var i = 0; i < vertCount; i++) {
      final x = prim.positions[i * 3].toStringAsFixed(6);
      final y = prim.positions[i * 3 + 1].toStringAsFixed(6);
      final z = prim.positions[i * 3 + 2].toStringAsFixed(6);
      buffer.writeln('v $x $y $z');
    }

    final triCount = prim.indices.length ~/ 3;
    for (var i = 0; i < triCount; i++) {
      final i0 = prim.indices[i * 3] + vertOffset;
      final i1 = prim.indices[i * 3 + 1] + vertOffset;
      final i2 = prim.indices[i * 3 + 2] + vertOffset;
      buffer.writeln('f $i0 $i1 $i2');
    }

    vertOffset += vertCount;
    buffer.writeln();
  }

  return buffer.toString();
}

/// Export a baked [Scene] to a Wavefront OBJ file at [path].
void exportObj(Scene scene, String path) {
  final text = toObj(scene);
  final file = File(path);
  final dir = file.parent;
  if (!dir.existsSync()) {
    dir.createSync(recursive: true);
  }
  file.writeAsStringSync(text);
}
